### PR TITLE
Add torch.utils.checkpoint.register_determinism_check

### PR DESCRIPTION
**Pull request by JaynouOliver**
Fixes #162980

## Summary

The `determinism_check` argument of `torch.utils.checkpoint.checkpoint` looks up its metadata function in the module-private dict `_allowed_determinism_checks_to_fns`, so users who want a custom check (e.g. one that also compares strides, or ignores devices) have to monkey-patch the dict. As @soulitzer said in the issue: "We'd accept a PR to add this."

This adds a public `register_determinism_check(name, metadata_fn)`:

- validates `metadata_fn` is callable and `name` isn't already registered (including the built-ins `"default"` and `"none"`),
- documents the contract (applied to every saved tensor on forward and recompute; results compared with `==`; mismatch raises `CheckpointError`),
- exports it in `__all__` and the `torch.utils.checkpoint` docs page, and updates the two `determinism_check` docstrings that said "these are the only two supported values".

**API scope note:** I deliberately kept the surface minimal (register only, no unregister/list) — the unknown-name error message already lists registered names dynamically, and unregistering has no clear use outside tests. Happy to extend if reviewers prefer the fuller trio.

## Test Plan

New `TestRegisterDeterminismCheck` in `test/test_utils.py`:

- registered check is actually invoked by `checkpoint(..., use_reentrant=False, determinism_check=<name>)` and gradients flow,
- a check returning different metadata between forward and recompute raises `CheckpointError` ("Recomputed values ... have different metadata"),
- duplicate / built-in names raise `ValueError`; non-callable raises `TypeError`,
- unknown `determinism_check` error message lists registered names.

```
$ python -m pytest test/test_utils.py::TestRegisterDeterminismCheck -v
5 passed
$ python -m pytest test/test_utils.py::TestCheckpoint -q
13 passed, 10 subtests passed
```

Verified on macOS (CPU) against a nightly wheel with the same patch applied.

Authored with the assistance of an AI assistant (Claude Code).

cc @soulitzer

### CODE REVIEW

**Comment by JaynouOliver** [comment]
Hi @albanD , @soulitzer can you please have a look?
